radicle-seed/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+pub struct Node {
+    config: NodeConfig,
+}
+
+impl Node {
+    /// Create a new seed node.
+    pub fn new(config: NodeConfig) -> Result<Self, Error> {
+        Ok(Node { config })
+    }
+
+    /// Run the seed node. This function runs indefinitely until a fatal error
+    /// occurs.
+    pub async fn run(self) -> Result<(), Error> {
+        let paths = if let Some(root) = self.config.root {
+            paths::Paths::from_root(root)?
+        } else {
+            paths::Paths::new()?
+        };
+        let gossip_params = Default::default();
+        let seeds: Vec<(PeerId, SocketAddr)> = vec![];
+        let disco = discovery::Static::new(seeds);
+        let signer = Signer::new();
+        let config = PeerConfig {
+            signer,
+            paths,
+            listen_addr: self.config.listen_addr,
+            gossip_params,
+            disco,
+        };
+
+        let peer = config.try_into_peer().await?;
+
+        let (api, future) = peer.accept()?;
+        let mut events = api.protocol().subscribe().await;
+        let mode = self.config.mode;
+
+        // Spawn the background peer thread.
+        tokio::spawn(future);
+
+        // Track already-known URNs.
+        Node::initialize_tracker(&mode, &api).await?;
+
+        // Listen on gossip events. As soon as a peer announces something of interest,
+        // we check if we should track it.
+        while let Some(event) = events.next().await {
+            match event {
+                ProtocolEvent::Gossip(gossip::Info::Has(gossip::Has { provider, val })) => {
+                    let urn = &val.urn;
+                    let peer_id = &provider.peer_id;
+
+                    tracing::info!("Discovered new URN {} from peer {}", urn, peer_id);
+
+                    if mode.is_trackable(peer_id, urn) {
+                        // Attempt to track, but keep going if it fails.
+                        Node::track_project(&api, urn, &provider).ok();
+                    }
+                },
+                _ => {},
+            }
+        }
+        Ok(())
+    }
+
+    /// Attempt to track a project.
+    fn track_project(
+        api: &PeerApi<Signer>,
+        urn: &RadUrn,
+        peer_info: &PeerInfo<std::net::IpAddr>,
+    ) -> Result<(), Error> {
+        let peer_id = &peer_info.peer_id;
+        let url = urn.clone().into_rad_url(peer_id.clone());
+        let port = peer_info.advertised_info.listen_port;
+        let addr_hints = peer_info
+            .seen_addrs
+            .iter()
+            .map(|a: &std::net::IpAddr| (*a, port).into());
+
+        let result = api
+            .storage()
+            .clone_repo::<project::ProjectInfo, _>(url, addr_hints)
+            .and_then(|_| api.storage().track(urn, &peer_id));
+
+        match &result {
+            Ok(()) => {
+                tracing::info!("Successfully tracked project {} from peer {}", urn, peer_id,);
+            },
+            Err(err) => {
+                tracing::debug!(
```

Comment:
Why not `warn!`?

Reply:
Because it's fairly normal for this to fail. A `warn` would imply that something could be wrong with the system and it needs to be checked out. Maybe @kim has a different opinion on this though.

Reply:
We will find out when things don't work as expected, and `RUST_LOG=debug` is too noisy :)